### ISSUE TITLE
Any way to disable prefetching ?

### ISSUE THREAD

**Issue by asmodehn** [celery/celery#2788]
My setup currently is :
- beat
- N workers ( concurrency 1 )
- 1 queue
- 1 very long running task

So it is a quite simple setup, the main difference with the usual celery setup being that the task is running for a long time ( from a few minutes to a few hours )

My current problem is : 
- worker 1 is free
- worker 2 is free

-> beat put a task in the queue
- worker 1 prefetch the task and starts executing it
- worker 1 is busy
- worker 2 is free

-> beat put a task in the queue
- worker 1 prefetch ( prefetch_count == 1 ) the task but doesn't run it.
- worker 2 is free

=> worker 2 should have gotten the task given that worker 1 was busy.

In my setup prefetching doesn't make much sense it seems. So is there any way to disable it ?
Setting it to 0 actually sets it to unlimited, it seems...

Another possible solution would be to have one queue per worker, and having a custom scheduler who knows who to send the task to, but this almost kills the point of using a distributed task management system...

Any hint ?

**Comment by kapilratnani** [comment]
I think this will [help](https://celery.readthedocs.org/en/latest/userguide/optimizing.html#prefork-pool-prefetch-settings)

**Comment by asmodehn** [comment]
Well, it didn't seem to make a difference for me...
I understand the -Ofair applies to the communication from worker to the pool processes ( same machine )

But I think my problem is on the worker itself, picking task from the queue while all pool processes (1) are already busy. I would like a different worker ( different machine ) to be able to grab that task instead, since that worker will likely not be able to execute that task for quite a long time.

**Comment by averwind** [comment]
just to be sure, you are setting CELERYD_PREFETCH_MULTIPLIER to 1 right and these are not tasks with ETA right?

**Comment by Thhhza** [comment]
I have the same setup and requirements as you do and it works for me. These are the settings I use:
CELERYD_CONCURRENCY = 1
CELERYD_PREFETCH_MULTIPLIER = 1

**Comment by asmodehn** [comment]
Oh some of my tasks are sent with an  an ETA indeed... I didnt expect this
to change the way the worker prefetch tasks... I ll check tasks without ETA
when i get some time.
Thank you for the information.

**Comment by asmodehn** [comment]
Here s my current configuration : 
I am using Celery 3.1.18.
CELERYD_PREFETCH_MULTIPLIER = 1
I pass concurrency 1 via command line argument
CELERY_ACKS_LATE = False since I don't want the task to be run a second time if it fail the first time

I ll keep investigating...

**Comment by asmodehn** [comment]
OK so it seems the PREFETCH behavior depends on the ACK behavior.
If ACKS_LATE = False : 
* Task 1 is prefetched 
* Task 1 is Ack'ed
* Task 2 is prefetched
* Task 1 is run ( Task 2 waiting a long time )

But if ACKS_LATE = True : 
* Task 1 is prefetched 
* Task 1 is run
* Task 1 is Ack'ed
* Task 2 is prefetched
* etc.
which is the behavior I wanted.

I guess I was confused because the docs mix the ACKS_LATE and the Retry behavior ?
http://celery.readthedocs.org/en/latest/faq.html#should-i-use-retry-or-acks-late

I think would be useful to have in the docs, a task flow diagram per configuration.
People have different uses of celery, and finding which configuration is the right one for your usecase is not obvious...

**Comment by ask** [comment]
The term 'prefetch' here comes from AMQP and is very confusing.

At no point is there prefetching involved, the QoS prefetch_count parameter is there to *limit* the number of messages held by the consumer at any time.

If you didn't have the prefetch count limit, the consumer would slurp in as many messages as quickly as possible.  The worker could avoid reading from the socket, but the kernel would still fill the kernel buffer with as many messages as that may hold.

So if you need it to process at most 1 message a time, you need to ack the message after processing it, not before (which ACKS_LATE will configure).

**Comment by ask** [comment]
Could you maybe open a separate issue for improving documentation?

**Comment by liormaynhypr** [comment]
> just to be sure, you are setting CELERYD_PREFETCH_MULTIPLIER to 1 right and these are not tasks with ETA right?

why does a task with an eta act differently?

**Comment by tayfun** [comment]
> > just to be sure, you are setting CELERYD_PREFETCH_MULTIPLIER to 1 right and these are not tasks with ETA right?
> 
> why does a task with an eta act differently?

See the note at https://docs.celeryproject.org/en/latest/userguide/configuration.html#worker-prefetch-multiplier:

> Tasks with ETA/countdown aren’t affected by prefetch limits.

So ETA tasks are prefetched without any limits but not run until the time to run them has come. I guess they will stay in the memory of the worker until the time comes for them to be run. Does this mean you shouldn't put a lot of tasks that will stay in the queue for long as they actually get sent to workers, ie. they don't stay in the queue until time comes for them to be sent. For example if you have 1000 messages with the ETA of 1 year later, they will be kept by worker for a year until it gets run.

In summary ETA tasks are supported by celery but only on RabbitMQ and not on SQS for example because SQS' other configuration is in conflict (visibility timeout will mean ETA task will be resent to a worker even though its time to run has not come).

**Comment by Thijsvandepoll** [comment]
Hi, I am facing a similar issue. It should be solved by setting `acks_late=True`, however, this setting causes the tasks to be replicated on newly added worker nodes, because it is not yet acknowledged. This is not desired. Is there any way to disable prefetching WITHOUT having `acks_late=True`? A more detailed reference to my problem is on StackOverflow: https://stackoverflow.com/questions/69987419/celery-prefetched-tasks-stuck-behind-other-tasks-on-ecs-cluster
